bin/agitator

```diff
@@ -30,6 +30,144 @@ Possible commands:
 EOF
 }
 
+function run_agitators() {
+  unset APP_HOSTS
+  export APP_HOSTS=$MANAGER_HOSTS
+  echo "Starting manager agitation as ${1} for ${APP_HOSTS}"
+  start_app_agitator "manager" "${AGTR_MASTER_KILL_SLEEP_TIME}" "${AGTR_MASTER_RESTART_SLEEP_TIME}" 1 1 > "${manager_log}" 2>&1 &
+  unset APP_HOSTS
+  export APP_HOSTS=$TSERVER_HOSTS
+  echo "Starting tserver agitation as ${1} for ${APP_HOSTS}"
+  start_app_agitator "tserver" "${AGTR_TSERVER_KILL_SLEEP_TIME}" "${AGTR_TSERVER_RESTART_SLEEP_TIME}" "${AGTR_TSERVER_MIN_KILL}" "${AGTR_TSERVER_MAX_KILL}" > "${tserver_log}" 2>&1 &
```

Comment:
These first three lines are duplicated with different parameters for the managers and the tservers. These can be passed as parameters to the `start_app_agitator` function instead.

Also, if at all possible, you should avoid passing around variables in and out of functions using global variables. This is especially important since you're backgrounding the calls and then immediately editing the value of the global variable, which could create concurrency issues. You can declare variables local to a function with the `local` keyword, as in `local x; x=$1` to assign the first argument to the function to the local variable named `x`.

Also, you don't need all the curly braces for scalar variables that aren't ambiguous in a string. They can add noise that affect readability.

```suggestion
  local agtr_user=$1
  start_app_agitator manager "$agtr_user" "$AGTR_MASTER_KILL_SLEEP_TIME" "$AGTR_MASTER_RESTART_SLEEP_TIME" 1 1 $MANAGER_HOSTS > "$manager_log" 2>&1 &
  start_app_agitator tserver "$agtr_user" "$AGTR_TSERVER_KILL_SLEEP_TIME" "$AGTR_TSERVER_RESTART_SLEEP_TIME" "$AGTR_TSERVER_MIN_KILL" "$AGTR_TSERVER_MAX_KILL" $TSERVER_HOSTS > "$tserver_log" 2>&1 &
```

Reply:
The `$APP_HOSTS` variable contains a string of values separated by a space so I was concerned they would expain as separated param if I passed them as a variable. I will test this out.

Reply:
They will, but that's okay in the example I gave, because I'm basically passing them as varargs. You can either pass them as a single, quoted value with space separated items, and then put that into an array later:

```bash
function printHosts() {
  local myarray=($1) # make the single argument turn into an array by not quoting so the spaces create separate items
  local x; for x in "${myarray[@]}"; do echo "$x"; done
}
myhosts="host1 host2 host3"
printHosts "$myhosts" # quotes pass it as single arg
```

Or, you can break them out like varargs (which is what I did above and in both the following examples):

Varargs example using an unquoted space-delimited scalar:

```bash
function printHosts() {
  local myarray=("$@") # put args into a new named array
  local x; for x in "${myarray[@]}"; do echo "$x"; done
}
myhosts="host1 host2 host3"
printHosts $myhosts # no quotes, so pass it as 3 args
```

Varargs example using an initial array where each element in the original array is passed as a separate argument:

```bash
function printHosts() {
  local myarray=("$@") # put args into a new named array
  local x; for x in "${myarray[@]}"; do echo "$x"; done
}
myhosts=("host1" "host2" "host3") # already an array
printHosts "${myhosts[@]}" # pass each element of the array as a separate arg
```

(Note: all of these examples are simplified, to avoid the `shift` that picks off the arguments before the varargs hosts.)

Reply:
Thanks for the tips. I already refactored the script to only use one function for all the agitators (except the hdfs one which still uses perl, that can be a follow on). I found a way to pass a variable nameref to a funtion to set the array for the hosts. Let me know what you think.